pandas/core/reshape/merge.py

```diff
@@ -1849,9 +1850,15 @@ def _right_outer_join(x, y, max_groups):
 
 def _factorize_keys(lk, rk, sort=True):
     # Some pre-processing for non-ndarray lk / rk
+    lk = extract_array(lk, extract_numpy=True)
+    rk = extract_array(rk, extract_numpy=True)
+
     if is_datetime64tz_dtype(lk) and is_datetime64tz_dtype(rk):
-        lk = getattr(lk, "_values", lk)._data
-        rk = getattr(rk, "_values", rk)._data
+        # Extract the ndarray (UTC-localized) values
+        # Note: we dont need the dtypes to match, as these can still be compared
```

Comment:
That does happen somewhere else?

Reply:
yes, in the EA-dtype check below we require matching dtypes